radix-engine/src/system/system.rs

```diff
@@ -2221,21 +2222,84 @@ where
     }
 
     #[trace_resources]
-    fn credit_cost_units(
-        &mut self,
-        vault_id: NodeId,
-        locked_fee: LiquidFungibleResource,
-        contingent: bool,
-    ) -> Result<LiquidFungibleResource, RuntimeError> {
-        self.api
+    fn start_credit_cost_units(&mut self, amount: Decimal) -> Result<bool, RuntimeError> {
+        let costing_enabled = self
+            .api
             .kernel_get_system()
             .modules
-            .apply_execution_cost(ExecutionCostingEntry::LockFee)?;
+            .enabled_modules
+            .contains(EnabledModules::COSTING);
 
+        // We do costing up front
         self.api
             .kernel_get_system()
             .modules
-            .credit_cost_units(vault_id, locked_fee, contingent)
+            .apply_execution_cost(ExecutionCostingEntry::LockFee)?;
+
+        let event_data = {
+            let lock_fee_event = LockFeeEvent { amount };
+            scrypto_encode(&lock_fee_event).unwrap()
+        };
+
+        // If costing is enabled, reserve event and pay for the event up front for the call to credit_cost_units()
+        // Otherwise, we just simulate the call
+        if costing_enabled {
```

Comment:
This is just following what was previously implemented. Would prefer to just skip the lock fee part but was worried that this would affect preview which may depend on these events?

Reply:
I believe disabling costing module is only for genesis and system transactions.

Slightly prefer to fail here if costing is not enabled, but maybe the PR implementation is fine.